File: tts/api-go/internal/app/voices.go
package app

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
)

type VoiceRegistry struct {
	manifestPath string
	manifestDir  string
	mu           sync.RWMutex
	manifestSize int64
	manifestTime int64
	rawVoices    []VoicePreset
	voices       map[string]VoicePreset
	ordered      []VoicePreset
}

func NewVoiceRegistry(manifestPath string) (*VoiceRegistry, error) {
	registry := &VoiceRegistry{
		manifestPath: manifestPath,
		manifestDir:  filepath.Dir(manifestPath),
	}
	if err := registry.Refresh(); err != nil {
		return nil, err
	}
	return registry, nil
}

func (r *VoiceRegistry) Refresh() error {
	manifestInfo, err := os.Stat(r.manifestPath)
	if err != nil {
		return fmt.Errorf("stat voice manifest: %w", err)
	}

	data, err := os.ReadFile(r.manifestPath)
	if err != nil {
		return fmt.Errorf("read voice manifest: %w", err)
	}

	var manifest VoiceManifestFile
	if err := json.Unmarshal(data, &manifest); err != nil {
		return fmt.Errorf("decode voice manifest: %w", err)
	}

	r.mu.Lock()
	r.manifestSize = manifestInfo.Size()
	r.manifestTime = manifestInfo.ModTime().UnixNano()
	r.rawVoices = append([]VoicePreset(nil), manifest.Voices...)
	r.rebuildLocked()
	r.mu.Unlock()
	return nil
}

func (r *VoiceRegistry) RefreshIfChanged() error {
	manifestInfo, err := os.Stat(r.manifestPath)
	if err != nil {
		return fmt.Errorf("stat voice manifest: %w", err)
	}

	r.mu.RLock()
	unchanged := len(r.rawVoices) > 0 && manifestInfo.Size() == r.manifestSize && manifestInfo.ModTime().UnixNano() == r.manifestTime
	r.mu.RUnlock()
	if !unchanged {
		data, err := os.ReadFile(r.manifestPath)
		if err != nil {
			return fmt.Errorf("read voice manifest: %w", err)
		}

		var manifest VoiceManifestFile
		if err := json.Unmarshal(data, &manifest); err != nil {
			return fmt.Errorf("decode voice manifest: %w", err)
		}

		r.mu.Lock()
		r.manifestSize = manifestInfo.Size()
		r.manifestTime = manifestInfo.ModTime().UnixNano()
		r.rawVoices = append([]VoicePreset(nil), manifest.Voices...)
		r.rebuildLocked()
		r.mu.Unlock()
		return nil
	}

	r.mu.Lock()
	r.rebuildLocked()
	r.mu.Unlock()
	return nil
}

func (r *VoiceRegistry) rebuildLocked() {
	voices := make(map[string]VoicePreset, len(r.rawVoices))
	ordered := make([]VoicePreset, 0, len(r.rawVoices))
	for _, voice := range r.rawVoices {
		resolved := voice
		switch voice.Source {
		case "builtin":
			if voice.EngineVoiceName != "" {
				resolved.Available = true
			} else {
				resolved.AvailabilityReason = "missing builtin engine voice mapping"
			}
		case "voice_pack":
			if voice.PackPromptAudio == "" {
				resolved.AvailabilityReason = "missing packaged prompt audio path"
				break
			}
			promptPath := filepath.Join(r.manifestDir, voice.PackPromptAudio)
			if _, err := os.Stat(promptPath); err == nil {
				resolved.Available = true
				resolved.Placeholder = false
			} else {
				resolved.AvailabilityReason = "packaged prompt audio not found"
			}
		default:
			resolved.AvailabilityReason = "unknown voice source"
		}

		voices[resolved.ID] = resolved
		ordered = append(ordered, resolved)
	}

	sort.SliceStable(ordered, func(i, j int) bool {
		return ordered[i].Label < ordered[j].Label
	})

	r.voices = voices
	r.ordered = ordered
}

func (r *VoiceRegistry) List() []VoicePreset {
	r.mu.RLock()
	defer r.mu.RUnlock()
	voices := make([]VoicePreset, len(r.ordered))
	copy(voices, r.ordered)
	return voices
}

func (r *VoiceRegistry) ResolveAvailable(id string) (VoicePreset, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	voice, ok := r.voices[id]
	if !ok {
		return VoicePreset{}, fmt.Errorf("voice preset %q not found", id)
	}
	if !voice.Available {
		reason := voice.AvailabilityReason
		if reason == "" {
			reason = "voice preset is unavailable"
		}
		return VoicePreset{}, errors.New(reason)
	}
	return voice, nil
}

func (r *VoiceRegistry) Readiness(requireCustomVoicePack bool) (bool, string) {
	voices := r.List()
	availableCount := 0
	for _, voice := range voices {
		if voice.Available {
			availableCount++
		}
	}
	if availableCount == 0 {
		return false, "no voice presets are currently available"
	}

	if requireCustomVoicePack {
		for _, voice := range voices {
			if voice.Source == "voice_pack" && !voice.Available {
				return false, fmt.Sprintf("custom voice pack %q is not packaged yet", voice.ID)
			}
		}
	}

	return true, "ok"
}

File: tts/worker/voices.py
from __future__ import annotations

import json
from pathlib import Path

try:  # pragma: no cover - import path compatibility
    from .errors import FatalJobError
    from .models import Job, ResolvedVoice, VoicePreset
except ImportError:  # pragma: no cover - direct script execution
    from errors import FatalJobError
    from models import Job, ResolvedVoice, VoicePreset


class VoiceRegistry:
    def __init__(self, manifest_path: Path) -> None:
        self.manifest_path = manifest_path.expanduser().resolve()
        self.manifest_dir = self.manifest_path.parent
        self._voices: dict[str, VoicePreset] = {}
        self._manifest_mtime_ns = 0
        self._manifest_size = -1
        self.refresh()

    def refresh(self) -> None:
        stat = self.manifest_path.stat()
        payload = json.loads(self.manifest_path.read_text(encoding="utf-8"))
        voices: dict[str, VoicePreset] = {}
        for raw_voice in payload.get("voices", []):
            voice = VoicePreset.from_dict(raw_voice)
            available = False
            reason = ""
            placeholder = voice.placeholder
            if voice.source == "builtin":
                available = bool(voice.engine_voice_name)
                if not available:
                    reason = "missing builtin engine voice mapping"
            elif voice.source == "voice_pack":
                if not voice.pack_prompt_audio:
                    reason = "missing packaged prompt audio path"
                else:
                    prompt_audio_path = (self.manifest_dir / voice.pack_prompt_audio).resolve()
                    if prompt_audio_path.is_file():
                        available = True
                        placeholder = False
                    else:
                        reason = "packaged prompt audio not found"
            else:
                reason = "unknown voice source"

            voices[voice.id] = VoicePreset(
                id=voice.id,
                label=voice.label,
                tone=voice.tone,
                source=voice.source,
                engine_voice_name=voice.engine_voice_name,
                pack_prompt_audio=voice.pack_prompt_audio,
                preview_path=voice.preview_path,
                placeholder=placeholder,
                available=available,
                availability_reason=reason,
            )
        self._voices = voices
        self._manifest_mtime_ns = getattr(stat, "st_mtime_ns", int(stat.st_mtime * 1_000_000_000))
        self._manifest_size = stat.st_size

    def refresh_if_changed(self) -> None:
        stat = self.manifest_path.stat()
        mtime_ns = getattr(stat, "st_mtime_ns", int(stat.st_mtime * 1_000_000_000))
        if mtime_ns == self._manifest_mtime_ns and stat.st_size == self._manifest_size and self._voices:
            return
        self.refresh()

    def resolve_for_job(self, job: Job) -> ResolvedVoice:
        self.refresh_if_changed()
        preset = self._voices.get(job.voice_preset_id)
        source = (job.voice_source or (preset.source if preset else "")).strip()
        if source == "builtin":
            engine_voice_name = (job.voice_engine_name or (preset.engine_voice_name if preset else "")).strip()
            if not engine_voice_name:
                raise FatalJobError(f"voice preset {job.voice_preset_id!r} is missing builtin engineVoiceName")
            return ResolvedVoice(source="builtin", engine_voice_name=engine_voice_name)

        if source == "voice_pack":
            if preset is None:
                raise FatalJobError(f"voice preset {job.voice_preset_id!r} was not found in the voice manifest")
            if not preset.pack_prompt_audio:
                raise FatalJobError(f"voice preset {job.voice_preset_id!r} is missing packPromptAudio")
            prompt_audio_path = (self.manifest_dir / preset.pack_prompt_audio).resolve()
            if not prompt_audio_path.is_file():
                raise FatalJobError(
                    f"voice preset {job.voice_preset_id!r} prompt audio not found: {prompt_audio_path}"
                )
            return ResolvedVoice(source="voice_pack", prompt_audio_path=prompt_audio_path)

        if preset is not None and preset.source == "builtin" and preset.engine_voice_name:
            return ResolvedVoice(source="builtin", engine_voice_name=preset.engine_voice_name)

        raise FatalJobError(
            f"voice preset {job.voice_preset_id!r} cannot be resolved for source {source or '<empty>'}"
        )
